httpclient5-cache/src/main/java/org/apache/hc/client5/http/impl/cache/AsyncCachingExec.java

```diff
@@ -498,22 +502,24 @@ public void cancelled() {
                 storeRequestIfModifiedSinceFor304Response(request, backendResponse);
             } else {
                 LOG.debug("Backend response is not cacheable");
-                responseCache.flushCacheEntriesFor(target, request, new FutureCallback<Boolean>() {
+                if (!Method.isSafe(request.getMethod())) {
```

Comment:
Perhaps we could consider delegating this check (`!Method.isSafe(request.getMethod())`) inside the method of the class `HttpAsyncCache` and `BasicHttpCache` itself. This way, the method becomes more resilient and we ensure the check is always performed when needed, reducing the potential for errors.

Reply:
@arturobernalg I am not sure. I do not think this is a cache storage layer decision whether a method is considered safe or not. But let's discuss it when I finish refactoring the cache eviction logic.